src/arch/io/network.cc

```diff
@@ -621,6 +621,430 @@ void linux_tcp_conn_t::on_event(int /* events */) {
+        return;
+    }
+
+    // Loop for retrying if the underlying socket would block and to retry on
+    // partial writes.
+    while (size > 0) {
+        ERR_clear_error();
+
+        int ret = SSL_write(conn, buffer, size);
+
+        if (ret > 0) {
+            // Operation successful, returns number of bytes read.
+            rassert(ret <= size);
+            size -= ret;
+
+            // Slide down the buffer.
+            buffer = reinterpret_cast<const void *>(reinterpret_cast<const char *>(buffer) + ret);
+
+            if (write_perfmon) write_perfmon->record(ret);
+
+            // Go around the loop again if there is more data to write.
+            continue;
+        }
+
+        int err_code = SSL_get_error(conn, ret);
+        unsigned long io_err = 0;
+
+        switch (err_code) {
+        case SSL_ERROR_ZERO_RETURN:
+            // Indicates that the peer has sent the "close notify" alert. The
+            // shutdown state is currently SSL_RECEIVED_SHUTDOWN. We must now
+            // send our "close notify" alert.
+            shutdown();
+            return;
+        case SSL_ERROR_WANT_READ:
+            /* There's no data available right now, so we must wait for a
+            notification from the epoll queue, or for an order to shut down. */
+            {
+                linux_event_watcher_t::watch_t watch(get_event_watcher(), poll_event_in);
+                wait_any_t waiter(&watch, &closed);
+                waiter.wait_lazily_unordered();
+            }
+            if (closed.is_pulsed()) {
+                /* We were closed for whatever reason. Something else has
+                already called on_shutdown(). In fact, we were
+                probably signalled by on_shutdown(). */
+                throw tcp_conn_read_closed_exc_t();
+            }
+            /* Go around the loop and try to read again */
+            continue;
+        case SSL_ERROR_WANT_WRITE:
+            /* Though we are reading, a TLS renegatiotain may occur at any time
+            requiring a write. Wait for a notification from the event queue, or
+            for an order to shut down */
+            {
+                linux_event_watcher_t::watch_t watch(get_event_watcher(), poll_event_out);
+                wait_any_t waiter(&watch, &closed);
+                waiter.wait_lazily_unordered();
+            }
+            if (closed.is_pulsed()) {
+                /* We were closed for whatever reason. Whatever signalled
+                us has already called on_shutdown(). */
+                throw tcp_conn_read_closed_exc_t();
+            }
+            /* Go around the loop and try to write again */
+            continue;
+        case SSL_ERROR_SYSCALL:
+            // An error with the underlying IO.
+            io_err = ERR_get_error();
+            if (io_err != 0) {
+                // The Openssl error queue has data on the error.
+                on_shutdown();
```

Comment:
I don't understand why it's safe to call `on_shutdown()` here (and in the couple cases below). How do we know that the connection was already shut down?
Should we call `shutdown()` instead?